# 1d Convolutional Neural Networks

In [0]:

# Imports
import numpy as np
import os
# from utils.utilities import *
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
import tensorflow as tf
import pandas as pd
from keras.models import Sequential
from keras.layers.core import Dense
import tensorflow as tf
from tensorflow import keras
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense, Dropout, Flatten
from keras.layers import Conv1D, MaxPooling1D
from keras.layers.convolutional import Conv1D
from keras import backend as K

### About the Data

Data is from people doing one of these activities (WALKING, WALKING_UPSTAIRS, WALKING_DOWNSTAIRS, SITTING, STANDING, LAYING) <br><br>
The data is from a smartphone and provides the body acceleration, general acceleration and a gyroscope reading <br><br>
Data is from: https://archive.ics.uci.edu/ml/datasets/human+activity+recognition+using+smartphones 

### Read in data from functions provided below

In [0]:
X_train, labels_train, list_ch_train = read_data_train() # train
X_test, labels_test, list_ch_test = read_data_test()

In [94]:
print(X_train.shape)

(7352, 128, 9)



#### Explaining the shape of the data 
There are 7,352 data points that are classified as doing one of those activities <br>
There are 128 time steps <br>
There are 9 values for each time step for each data point for the x,y,z values of the body 
acceleration, general acceleration and gyrscope reading. <br>

### Use train_test_split to create training and validation data and labels

In [95]:
X_train, X_val, y_train, y_val = train_test_split(X_train, labels_train, test_size=0.25, random_state=0)
X_train

array([[[ 1.481724e-01,  1.046317e-01,  3.250692e-04, ...,
          1.159696e+00, -2.713632e-02,  8.817388e-02],
        [ 1.435533e-01,  2.076716e-01, -9.300516e-02, ...,
          1.155066e+00,  7.606756e-02, -5.106673e-03],
        [ 1.811939e-01,  1.544515e-01, -1.273095e-01, ...,
          1.192768e+00,  2.300220e-02, -3.937873e-02],
        ...,
        [-1.275830e-01,  6.643515e-02,  8.251871e-02, ...,
          8.905337e-01, -5.582116e-02,  1.741138e-01],
        [-1.147077e-01,  8.737952e-02,  1.391429e-01, ...,
          9.034618e-01, -3.480743e-02,  2.304481e-01],
        [-1.520196e-01,  7.907708e-02,  1.563400e-01, ...,
          8.661492e-01, -4.303240e-02,  2.473652e-01]],

       [[-4.666965e-02,  4.995205e-02,  6.789561e-02, ...,
          8.617538e-01,  2.953420e-01,  4.560750e-01],
        [-4.183416e-02,  5.042665e-02,  6.845513e-02, ...,
          8.662995e-01,  2.963811e-01,  4.568254e-01],
        [-4.006963e-02,  4.790394e-02,  6.308372e-02, ...,
          8.67

One hot coding is when each row represents one label and it the nonzero column represents what the label is<br>
 [0,0,0,0,0,1] <br>
           ^ This label is LAYING, which was previously represented with a 5
           <br><br>

### One hot encoding functions are provided at the bottom of this notebook <br> Use these functions to convert the training, validation and testing labels

In [0]:
y_train = one_hot(y_train)
y_test = one_hot(labels_test)
y_val = one_hot(y_val)

### Define your Keras model <br>
You will want to use

See for examples on creating Keras modelshttps://github.com/keras-team/keras/blob/master/examples/mnist_cnn.py

In [97]:
# fill in with your code below
model = keras.Sequential()
model.add(keras.layers.Conv1D(filters=18, kernel_size=3, strides=1, padding='same', activation='relu', input_shape=X_train[0].shape))
model.add(keras.layers.MaxPooling1D(pool_size=2, padding='valid'))
model.add(keras.layers.Flatten())

model.add(keras.layers.Dense(y_train[0].shape[0], activation='relu'))
print(model.summary())  # view model
    
    


_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv1d_8 (Conv1D)            (None, 128, 18)           504       
_________________________________________________________________
max_pooling1d_8 (MaxPooling1 (None, 64, 18)            0         
_________________________________________________________________
flatten_1 (Flatten)          (None, 1152)              0         
_________________________________________________________________
dense_5 (Dense)              (None, 6)                 6918      
Total params: 7,422
Trainable params: 7,422
Non-trainable params: 0
_________________________________________________________________
None


### Compile your model

In [0]:
lr =  .0008# choose a learning rate, this will be a good paramter to tune
model.compile( optimizer=tf.train.AdamOptimizer(lr), loss=keras.losses.categorical_crossentropy, metrics = ['accuracy'] )

### Fit your model

In [99]:
history = model.fit(X_train,
                    y_train,
                    epochs= 100,
                    batch_size= 600,
                    validation_data=(X_val, y_val),
                    verbose=1 
                    )

Train on 5514 samples, validate on 1838 samples
Epoch 1/100
5514/5514 [==============================] - 1s 139us/step - loss: nan - acc: 0.1563 - val_loss: nan - val_acc: 0.1676
Epoch 2/100
5514/5514 [==============================] - 1s 96us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 3/100
5514/5514 [==============================] - 0s 88us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 4/100
5514/5514 [==============================] - 1s 91us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 5/100
5514/5514 [==============================] - 0s 87us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 6/100
5514/5514 [==============================] - 0s 89us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 7/100
5514/5514 [==============================] - 0s 89us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 8/100
5514/5514 [==================

5514/5514 [==============================] - 0s 86us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 16/100
5514/5514 [==============================] - 0s 86us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 17/100
5514/5514 [==============================] - 0s 88us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 18/100
5514/5514 [==============================] - 0s 89us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 19/100
5514/5514 [==============================] - 0s 87us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 20/100
5514/5514 [==============================] - 0s 87us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 21/100
5514/5514 [==============================] - 0s 87us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 22/100
5514/5514 [==============================] - 1s 92us/step - loss: nan - acc: 0.1665

5514/5514 [==============================] - 0s 88us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 30/100
5514/5514 [==============================] - 0s 86us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 31/100
5514/5514 [==============================] - 0s 87us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 32/100
5514/5514 [==============================] - 0s 88us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 33/100
5514/5514 [==============================] - 0s 89us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 34/100
5514/5514 [==============================] - 0s 88us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 35/100
5514/5514 [==============================] - 0s 89us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 36/100
5514/5514 [==============================] - 0s 88us/step - loss: nan - acc: 0.1665

5514/5514 [==============================] - 0s 90us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 44/100
5514/5514 [==============================] - 0s 88us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 45/100
5514/5514 [==============================] - 0s 88us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 46/100
5514/5514 [==============================] - 0s 87us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 47/100
5514/5514 [==============================] - 0s 90us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 48/100
5514/5514 [==============================] - 0s 90us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 49/100
5514/5514 [==============================] - 0s 88us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 50/100
5514/5514 [==============================] - 0s 89us/step - loss: nan - acc: 0.1665

5514/5514 [==============================] - 0s 87us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 58/100
5514/5514 [==============================] - 0s 88us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 59/100
5514/5514 [==============================] - 0s 90us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 60/100
5514/5514 [==============================] - 0s 89us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 61/100
5514/5514 [==============================] - 0s 88us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 62/100
5514/5514 [==============================] - 0s 87us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 63/100
5514/5514 [==============================] - 0s 88us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 64/100
5514/5514 [==============================] - 0s 89us/step - loss: nan - acc: 0.1665

5514/5514 [==============================] - 1s 91us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 72/100
5514/5514 [==============================] - 0s 88us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 73/100
5514/5514 [==============================] - 0s 88us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 74/100
5514/5514 [==============================] - 0s 86us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 75/100
5514/5514 [==============================] - 0s 88us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 76/100
5514/5514 [==============================] - 0s 87us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 77/100
5514/5514 [==============================] - 0s 89us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 78/100
5514/5514 [==============================] - 0s 90us/step - loss: nan - acc: 0.1665

5514/5514 [==============================] - 0s 88us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 86/100
5514/5514 [==============================] - 0s 89us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 87/100
5514/5514 [==============================] - 0s 87us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 88/100
5514/5514 [==============================] - 0s 88us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 89/100
5514/5514 [==============================] - 0s 88us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 90/100
5514/5514 [==============================] - 1s 91us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 91/100
5514/5514 [==============================] - 0s 88us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 92/100
5514/5514 [==============================] - 0s 88us/step - loss: nan - acc: 0.1665

5514/5514 [==============================] - 0s 89us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676
Epoch 100/100
5514/5514 [==============================] - 0s 87us/step - loss: nan - acc: 0.1665 - val_loss: nan - val_acc: 0.1676


### Code below prints the test loss and accuracy <br>
### Before you print your accuracy and loss, play around with the parameters!!

In [101]:
score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

('Test loss:', nan)
('Test accuracy:', 0.168306752629793)


### Provided Functions Below

In [0]:
def read_data_train():
  """ Read data """

  # Fixed params
  n_class = 6
  n_steps = 128

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/y_train.txt'
  labels = pd.read_csv(label_path, header = None)

  list_of_channels = ['body_acc_x', 'body_acc_y', 'body_acc_z', 'body_gyro_x',
  'body_gyro_y', 'body_gyro_z', 'total_acc_x', 'total_acc_y', 'total_acc_z']

  X = np.zeros((len(labels), n_steps, len(list_of_channels)))

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_acc_x_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,0] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_acc_y_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,1] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_acc_z_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,2] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_gyro_x_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,3] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_gyro_y_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,4] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_gyro_z_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,5] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/total_acc_x_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,6] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/total_acc_y_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,7] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/total_acc_z_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,8] = dat_.as_matrix()


  # Return 
  return X, labels[0].values, list_of_channels

In [0]:
def read_data_test():
  """ Read data """

  # Fixed params
  n_class = 6
  n_steps = 128

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/y_test.txt'
  labels = pd.read_csv(label_path, header = None)

  list_of_channels = ['body_acc_x', 'body_acc_y', 'body_acc_z', 'body_gyro_x',
  'body_gyro_y', 'body_gyro_z', 'total_acc_x', 'total_acc_y', 'total_acc_z']

  X = np.zeros((len(labels), n_steps, len(list_of_channels)))

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_acc_x_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,0] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_acc_y_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,1] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_acc_z_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,2] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_gyro_x_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,3] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_gyro_y_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,4] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_gyro_z_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,5] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/total_acc_x_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,6] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/total_acc_y_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,7] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/total_acc_z_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,8] = dat_.as_matrix()


  # Return 
  return X, labels[0].values, list_of_channels

In [0]:
def one_hot(labels, n_class = 6):
	""" One-hot encoding """
	expansion = np.eye(n_class)
	y = expansion[:, labels-1].T
	assert y.shape[1] == n_class, "Wrong number of labels!"

	return y

In [0]:
def get_batches(X, y, batch_size = 100):
	""" Return a generator for batches """
	n_batches = len(X) // batch_size
	X, y = X[:n_batches*batch_size], y[:n_batches*batch_size]

	# Loop over batches and yield
	for b in range(0, len(X), batch_size):
		yield X[b:b+batch_size], y[b:b+batch_size]